src/Memoize.jl

```diff
@@ -1,89 +1,196 @@
 module Memoize
-using MacroTools: isexpr, combinedef, namify, splitarg, splitdef
-export @memoize, memoize_cache
+using MacroTools: isexpr, combinearg, combinedef, namify, splitarg, splitdef, @capture
+export @memoize, memories, memory
 
-cache_name(f) = Symbol("##", f, "_memoized_cache")
+# I would call which($sig) but it's only on 1.6 I think
+function _which(tt, world = typemax(UInt))
+    meth = ccall(:jl_gf_invoke_lookup, Any, (Any, UInt), tt, world)
+    if meth !== nothing
+        if meth isa Method
+            return meth::Method
+        else
+            meth = meth.func
+            return meth::Method
+        end
+    end
+end
+
+const _brain = Dict()
```

Comment:
A global like this is problematic because of precompilation. Eg. if you have a package with

```julia
module X
@memoize foo(x) = x
end
```

I believe you'll find that after `using Memoize, X`, `Memoize.brain` is empty.